src/Build/Logging/TerminalLogger/TerminalLogger.cs

```diff
@@ -684,19 +689,43 @@ private void ProjectFinished(object sender, ProjectFinishedEventArgs e)
+                            // some options:
+                            // * the raw, full output path from the MSBuild logic (OutputPath property)
+                            // * the output path relative to the initial working directory, if it is under it
+                            // * the output path relative to the source root, if it is under it
+
+                            // full path fallback
+                            var projectDisplayPathSpan = outputPathSpan;
+                            var workingDirectorySpan = _initialWorkingDirectory.AsSpan();
+                            // under working dir case
+                            if (outputPathSpan.StartsWith(workingDirectorySpan, FileUtilities.PathComparison))
+                            {
+                                if (outputPathSpan.Length > workingDirectorySpan.Length
+                                    && (outputPathSpan[workingDirectorySpan.Length] == Path.DirectorySeparatorChar
+                                        || outputPathSpan[workingDirectorySpan.Length] == Path.AltDirectorySeparatorChar))
+                                {
+                                    projectDisplayPathSpan = outputPathSpan.Slice(workingDirectorySpan.Length + 1);
+                                }
+                            }
+                            // under source root case
+                            else if (project.SourceRoot is { Span: var sourceRootSpan } )
                             {
-                                if (outputPathSpan.Length > _initialWorkingDirectory.Length
-                                    && (outputPathSpan[_initialWorkingDirectory.Length] == Path.DirectorySeparatorChar
-                                        || outputPathSpan[_initialWorkingDirectory.Length] == Path.AltDirectorySeparatorChar))
+                                var relativePathFromWorkingDirToSourceRoot = Path.GetRelativePath(workingDirectorySpan.ToString(), sourceRootSpan.ToString()).AsSpan();
+                                if (outputPathSpan.StartsWith(sourceRootSpan, FileUtilities.PathComparison))
                                 {
-                                    outputPathSpan = outputPathSpan.Slice(_initialWorkingDirectory.Length + 1);
+                                    if (outputPathSpan.Length > sourceRootSpan.Length
+                                        // offsets are -1 here compared to above for ***reasons***
+                                        && (outputPathSpan[sourceRootSpan.Length - 1] == Path.DirectorySeparatorChar
+                                            || outputPathSpan[sourceRootSpan.Length - 1] == Path.AltDirectorySeparatorChar))
+                                    {
+                                        projectDisplayPathSpan = Path.Combine(relativePathFromWorkingDirToSourceRoot.ToString(), outputPathSpan.Slice(sourceRootSpan.Length).ToString()).AsSpan();
+                                    }
```

Comment:
And here is where we can then use the source root information to render a nice relative path for users like we do for paths that are in the same direct project tree. This will make .NET SDK artifacts layouts look much nicer for example.